### PR TITLE
fix issue 485: panic error after not finding matching VM in k3s

### PR DESCRIPTION
**Pull request by nicolehanjing**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes https://github.com/kubernetes/cloud-provider-vsphere/issues/485

**Special notes for your reviewer**:

**Release note**:

```release-note
fix issue 485: panic error after not finding matching VM in k3s
```

### CODE REVIEW

**Review by lubronzhan — APPROVED** [review]
LGTM

File: releases/v1.20/vsphere-cloud-controller-manager.yaml [hunk 1]
```diff
@@ -0,0 +1,253 @@
+---
```
Comment (andrewsykim) [inline releases/v1.20/vsphere-cloud-controller-manager.yaml:1]: Can we split the manifest update into a separate PR so the panic is easier to cherry-pick to older release?

Comment (nicolehanjing) [reply releases/v1.20/vsphere-cloud-controller-manager.yaml:1]: gotcha, created another PR: https://github.com/kubernetes/cloud-provider-vsphere/pull/487/files for adding manifests

**Review by lubronzhan — APPROVED** [review]
lgtm

File: pkg/cloudprovider/vsphere/instances.go [hunk 2]
```diff
@@ -130,14 +130,20 @@ func (i *instances) InstanceID(ctx context.Context, nodeName types.NodeName) (st
 // InstanceType returns the type of the instance identified by name.
 func (i *instances) InstanceType(ctx context.Context, name types.NodeName) (string, error) {
 	klog.V(4).Info("instances.InstanceType() called")
-	return i.nodeManager.nodeNameMap[string(name)].NodeType, nil
+	if nodeInfo, ok := i.nodeManager.nodeNameMap[string(name)]; ok {
+		return nodeInfo.NodeType, nil
+	}
+	return "", nil
```
Comment (andrewsykim) [inline pkg/cloudprovider/vsphere/instances.go:136]: Is it appropriate to return a nil error here or should we error if the object doesn't exist in cache?

File: pkg/cloudprovider/vsphere/instances.go [hunk 3]
```diff
@@ -130,14 +131,20 @@ func (i *instances) InstanceID(ctx context.Context, nodeName types.NodeName) (st
 // InstanceType returns the type of the instance identified by name.
 func (i *instances) InstanceType(ctx context.Context, name types.NodeName) (string, error) {
 	klog.V(4).Info("instances.InstanceType() called")
-	return i.nodeManager.nodeNameMap[string(name)].NodeType, nil
+	if nodeInfo, ok := i.nodeManager.nodeNameMap[string(name)]; ok {
+		return nodeInfo.NodeType, nil
+	}
+	return "", fmt.Errorf("cannot find node with nodeName %s in nodeUUIDMap", name)
```
Comment (nicolehanjing) [inline pkg/cloudprovider/vsphere/instances.go:137]: @andrewsykim I added the error message, PTAL :)

File: docs/book/tutorials/deploying-cpi-with-k3s.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
+# Deploying the vSphere CPI using k3s
```
Comment (andrewsykim) [inline docs/book/tutorials/deploying-cpi-with-k3s.md:1]: this README should also be opened in a separate PR and commit

**Review by andrewsykim — COMMENTED** [review]
minor comment about README being in a separate commit or PR, otherwise LGTM!

File: docs/book/tutorials/deploying-cpi-with-k3s.md (hunk 4 above)
Comment (nicolehanjing) [reply docs/book/tutorials/deploying-cpi-with-k3s.md]: added the doc in another PR: https://github.com/kubernetes/cloud-provider-vsphere/pull/488

File: pkg/cloudprovider/vsphere/instances.go (hunk 3 above)
Comment (andrewsykim) [reply pkg/cloudprovider/vsphere/instances.go:137]: minor typo in this error message, should be nodeNameMap, not nodeUUIDMap

**Review by andrewsykim — COMMENTED** [review]
/lgtm
